File: Data Structures & Algorithms/buy-and-sell-crypto/submission-0.cpp
class Solution {
public:
   
 int maxProfit(vector<int>& prices) {

        int l = 0, r = 0 , ans = 0;

        while(r<prices.size()){

               ans = max(ans, prices[r] - prices[l]);
               if (prices[l] > prices[r])
                      l++;

               r++;
        }
        return ans;
 }

};

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-0.js
class Solution {
    /**
     * @param {number[]} prices
     * @return {number}
     */
    maxProfit(prices) {
        const memo = {} ;
        function dfs(i=0, isBuy = true){
           
            if (i >= prices.length) return 0;
            const key = `${i}:${isBuy}`;
            if (memo[key] !== undefined) return memo[key];
            let cooldown = dfs(i + 1, isBuy);
            let ans = 0;
            if(isBuy)
                memo[key] = Math.max(dfs(i + 1, false) - prices[i], cooldown);
            else 
            memo[key] = Math.max(dfs(i + 2, true) + prices[i], cooldown);
           
            return memo[key];
        }

        return dfs()

    }
}
